packages/slate-react/src/utils/scroll-to-selection.js

```diff
@@ -53,41 +55,101 @@ function scrollToSelection(selection) {
   if (!selection.anchorNode) return
 
   const window = getWindow(selection.anchorNode)
-  const scroller = findScrollContainer(selection.anchorNode)
-  const isWindow = scroller == window
+  const scroller = findScrollContainer(selection.anchorNode, window)
+  const isWindow =
+    scroller == window.document.body ||
+    scroller == window.document.documentElement
   const backward = isBackward(selection)
+
   const range = selection.getRangeAt(0)
-  const rect = range.getBoundingClientRect()
+  let selectionRect = range.getBoundingClientRect()
+
+  // range.getBoundingClientRect() returns 0s in Safari when range is collapsed.
+  // Expanding the range by 1 is a relatively effective workaround for vertical
+  // scroll, although horizontal may be off by 1 character.
+  // https://bugs.webkit.org/show_bug.cgi?id=138949
+  // https://bugs.chromium.org/p/chromium/issues/detail?id=435438
+  if (range.collapsed && selectionRect.top == 0 && selectionRect.height == 0) {
+    if (range.startOffset == 0) {
+      range.setEnd(range.endContainer, 1)
+    } else {
+      range.setStart(range.startContainer, range.startOffset - 1)
+    }
+
+    selectionRect = range.getBoundingClientRect()
+  }
+
   let width
   let height
   let yOffset
   let xOffset
+  let scrollerTop = 0
+  let scrollerLeft = 0
+  let scrollerBordersY = 0
+  let scrollerBordersX = 0
 
   if (isWindow) {
-    const { innerWidth, innerHeight, pageYOffset, pageXOffset } = scroller
+    const { innerWidth, innerHeight, pageYOffset, pageXOffset } = window
     width = innerWidth
     height = innerHeight
     yOffset = pageYOffset
     xOffset = pageXOffset
   } else {
     const { offsetWidth, offsetHeight, scrollTop, scrollLeft } = scroller
+    const {
+      borderTopWidth,
+      borderBottomWidth,
+      borderLeftWidth,
+      borderRightWidth,
+    } = window.getComputedStyle(scroller)
+
+    const scrollerBorderTop = parseInt(borderTopWidth, 10)
+    const scrollerBorderBottom = parseInt(borderBottomWidth, 10)
+    const scrollerBorderLeft = parseInt(borderLeftWidth, 10)
+    const scrollerBorderRight = parseInt(borderRightWidth, 10)
```

Comment:
If these aren't used anywhere else, can we just inline them in the variables below instead?

Reply:
Sure, updated.